Add garbage collection for text type design document (#143)
yorkie-team/yorkie: design/gc-for-text-type.md

---
title: gc-for-text-type
target-version: 0.1.1
---

# Garbage collection for Text Type

## Summary

Text and RichText use `RgaTreeSplit`. And when the text is removed, it marks a tombstone on the node corresponding to the text.  
We provide garbage collection to clean up nodes of text type data that have been removed.

### Goals

When all synchronization between clients is complete, the node marked tombstone is no longer needed.
As the number of nodes marked tombstone increases, unnecessary memory waste occurs.

> So it provides garbage collection to solve the problem of wasting memory.

Reviewer: I think this explanation fits better in the Summary. I think it is better than what is currently described in the Summary.

In this section, it would be better to write a specific goal rather than a description of the situation.

```
Implements garbage collection for text types, Text, and RichText. When
executing garbage collection of a document, text nodes marked with
tombstones are also deleted.
```

### Non-Goals

1. It does not directly control the garbage collection provided by the programming language. It simply removes the relationship of the nodes connected to the linked list and makes it a collection target.

> 2. It is not a function that can be called directly by the yorkie user.

Reviewer: It would be nice to explain this item in Proposal details along with the reason.

## Proposal details

### How dose it work?

If there is a node that is deleted when the text is modified, store the text element at `Root` instance.
At this time, it is stored inside `Document` provided to the user and inside `Document` used in internal logic.
That's because the two `Document` use separate memory addresses, and even the same text values have different memory addresses.

And when the client synchronizes, it calls `GarbageCollect()` inside the `ApplyChangePack()` method.
When `GarbageCollect()` is called, the node to be deleted internally is excluded from the linked list.
What you need to know at this time is that `RGATreeSplitNode` has `insPrev` and `insNext`.
Must also do the work to connect the relationship between `insPrev` and `insNext`.
> `insPrev` and `insNext` are used to remember other nodes connected to the insert.

> > For example, when `abc` is divided into `a`, `b`, `c`, it looks like this: [abc] divided to [a]<->[b]<->[c]

Reviewer: It seems that this section is still being written. It would be nice to explain it with a diagram later.

---
title: gc-for-text-type
target-version: 0.1.1
---

> # Garbage collection for Text Type

Reviewer 2: Garbage collection for Text Type => Garbage Collection for Text Type

## Summary

> When all synchronization between clients is complete, the node marked tombstones is no longer needed.

Reviewer: ```suggestion
When synchronization between clients is complete, the nodes marked as tombstones are no longer needed.
```

> As the number of nodes marked tombstone increases, unnecessary memory waste occurs.

Reviewer 2: marked tombstone => marked tombstones

> As the number of nodes marked tombstones increases, unnecessary memory waste occurs.

Reviewer: ```suggestion
Increasing the number of nodes marked as tombstones lead to unnecessary waste of memory.
```

> When synchronization between clients is complete, the nodes marked as tombstones are no longer needed.
> Increasing the number of nodes marked as tombstones lead to unnecessary waste of memory.
> So it provides garbage collection to solve the problem of wasting memory.

Reviewer: ```suggestion
We only implemented garbage collection for elements marked as tombstones to reduce unnecessary waste of memory. This feature implements a garbage collection for text nodes marked as tombstones that occur within elements of text type.
```

### Goals

Implements garbage collection for text types, Text, and RichText. When
executing garbage collection of a document, text nodes marked with
tombstones are also deleted.

### Non-Goals

1. It does not directly control the garbage collection provided by the programming language. It simply removes the relationship of the nodes connected to the linked list and makes it a collection target.

> 2. It is not a function that can be called directly by the yorkie user.

Reviewer: ```suggestion
2. It is not a feature that Yorkie users can use directly.
```

> ## Proposal details

Reviewer 2: Proposal details => Proposal Details

> ### How dose it work?

Reviewer 2: dose => does

#### Save changed text

> Text nodes that are deleted when editing text become tombstones and are stored in the `RGATreeSplit` instance.

Reviewer: ```suggestion
Text nodes that are deleted during editing are marked as tombstones and stored in the `RGATreeSplit` instance.
```

> And text elements with text nodes that are deleted during the editing process are stored in the `Root` instance.

Reviewer 2: And text elements => Additionally, text elements

> And `Root` instance is managed by `Document` instance.

Reviewer 2: And the `Root` instance is managed by the `Document` instance.

The important point when text elements are stored is that they must be stored in both the `Document` that is provided to the user and the `Document` that holds the original data.
That's because the two `Document` use separate memory addresses, and even the same text values have different memory addresses.

```go
// pkg/document/document.go

type Document struct {
	// doc is the original data of the actual document.
	doc *InternalDocument

	// clone is a copy of `doc` to be exposed to the user and is used to
	// protect `doc`.
	clone *json.Root
}
```

When garbage collection is enabled, both the original and copied data are checked to collect garbage.

#### Garbage collection run

> And when the client syncs, Yorkie's garbage collection runs. 

Reviewer 2: When the client syncs, Yorkie's garbage collection runs automatically

> >The important thing is that the user does not call it directly. If the user calls it directly, it may cause unintended behavior and cause problems for the entire program.

Reviewer 2: cause problems => lead to

>Currently, `Document.GarbageCollect` is exposed to users as a public method, so it may be misunderstood. We are aware of this issue and are working through issue [#125](https://github.com/yorkie-team/yorkie/issues/125).

> Cleans up tombstone marked nodes by traversing the modified text elements stored in the `Root` instance.

Reviewer 2: Cleans up => It cleans up
tombstone marked => tombstone-marked

Organized nodes are excluded from the linked list and are subject to garbage collection provided by the programming language.

```bash
# input 'abc'
 +-----+        +-----+        +-----+
 | 'a' |   ->   | 'b' |   ->   | 'c' |
 +-----+        +-----+        +-----+

# remove 'b'
              (tombstoned)
 +-----+        +-----+        +-----+
 | 'a' |   ->   | 'b' |   ->   | 'c' |
 +-----+        +-----+        +-----+

# after running gc
 +-----+        +-----+
 | 'a' |   ->   | 'c' |
 +-----+        +-----+

           (Collection target)
               (tomstone)    
                +-----+
                | 'b' |
                +-----+
```
>It only makes it a garbage collection target. If we use `runtime.GC()` directly, performance degradation will occur due to frequent `GC` calls.

What you need to know at this time is that `RGATreeSplitNode` has `insPrev` and `insNext`.

> Must also do the work to connect the relationship between `insPrev` and `insNext`.

Reviewer 2: Must also => It must also
